pkg/inttest/inttest.go

```diff
@@ -0,0 +1,315 @@
+// Package inttest enables writing of integration tests. Setup Docker containers for dependencies
+// like PostgreSQL, RabbitMQ and AWS S3 (using localstack). Every setup function ensures the
+// container is ready before returning, ensures resources are cleaned up after the tests are
+// finished and return a client ready to interact with the container.
+package inttest
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/go-redis/redis"
+	gnomockRedis "github.com/orlangure/gnomock/preset/redis"
+
+	amqp "github.com/rabbitmq/amqp091-go"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/dhis2-sre/im-manager/internal/handler"
+	"github.com/dhis2-sre/im-manager/internal/server"
+	"github.com/dhis2-sre/im-manager/pkg/storage"
+	"github.com/gin-gonic/gin"
+	_ "github.com/lib/pq" // postgres driver
+	"github.com/orlangure/gnomock"
+	"github.com/orlangure/gnomock/preset/localstack"
+
+	"github.com/dhis2-sre/im-manager/pkg/config"
+	"github.com/orlangure/gnomock/preset/postgres"
+	"github.com/orlangure/gnomock/preset/rabbitmq"
+
+	"github.com/stretchr/testify/require"
+	"gorm.io/gorm"
+)
+
+func SetupRedis(t *testing.T) *redis.Client {
+	container, err := gnomock.Start(gnomockRedis.Preset())
+	require.NoError(t, err, "failed to start Redis")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop Redis") })
+
+	client := &redis.Options{
+		Addr:     container.DefaultAddress(),
+		Password: "",
+		DB:       0,
+	}
+	return redis.NewClient(client)
+}
+
+// SetupDB creates a PostgreSQL container. Gorm is connected to the DB and runs the migrations.
+func SetupDB(t *testing.T) *gorm.DB {
+	t.Helper()
+
+	container, err := gnomock.Start(
+		postgres.Preset(
+			postgres.WithUser("im", "im"),
+			postgres.WithDatabase("test_im"),
+		),
+	)
+	require.NoError(t, err, "failed to start DB")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop DB") })
+
+	db, err := storage.NewDatabase(config.Postgresql{
+		Host:         container.Host,
+		Port:         container.DefaultPort(),
+		Username:     "im",
+		Password:     "im",
+		DatabaseName: "test_im",
+	})
+	require.NoError(t, err, "failed to setup DB")
+	return db
+}
+
+// SetupS3 creates an S3 container (using localstack) with all the buckets and files in given path.
+func SetupS3(t *testing.T, path string) *S3Client {
```

Comment:
The inttest provides helper clients that wrap methods of some client like an HTTP or AWS client. The idea is to call the original client and assert. I wrote tests first without creating such a client and then extracted the repetitive code like setup and assertions. We should not reuse clients of our production code as this would mean that our tests are not testing them.